### ISSUE TITLE
[Warning Handling] Avoid using 'tokenizers' before the fork if possible

### ISSUE THREAD

**Issue by ArtemisDicoTiar** [eubinecto/wisdomify#21]
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)

워닝 간단하게 해결할 수 있을 거 같다.
해보고 안되면 나중에 수정!

**Comment by ArtemisDicoTiar** [comment]
@eubinecto 이 워닝 어디서 어떤거 할때 발생하는 지 알려주면 내가 수정해볼게!

**Comment by eubinecto** [comment]
> @eubinecto 이 워닝 어디서 어떤거 할때 발생하는 지 알려주면 내가 수정해볼게!

```bash
python3 -m wisdomify.main.eval
```

eval script 실행하면 바로 reproduce 할 수 있을 것, 이렇게:
```
/Users/eubin/Desktop/Projects/Big/wisdomify/wisdomifyenv/bin/python /Users/eubin/Desktop/Projects/Big/wisdomify/wisdomify/main/eval.py
/Users/eubin/Desktop/Projects/Big/wisdomify/wisdomifyenv/lib/python3.8/site-packages/pytorch_lightning/metrics/__init__.py:43: LightningDeprecationWarning: `pytorch_lightning.metrics.*` module has been renamed to `torchmetrics.*` and split off to its own package (https://github.com/PyTorchLightning/metrics) since v1.3 and will be removed in v1.5
  rank_zero_deprecation(
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
/Users/eubin/Desktop/Projects/Big/wisdomify/wisdomifyenv/lib/python3.8/site-packages/pytorch_lightning/metrics/__init__.py:43: LightningDeprecationWarning: `pytorch_lightning.metrics.*` module has been renamed to `torchmetrics.*` and split off to its own package (https://github.com/PyTorchLightning/metrics) since v1.3 and will be removed in v1.5
  rank_zero_deprecation(
/Users/eubin/Desktop/Projects/Big/wisdomify/wisdomifyenv/lib/python3.8/site-packages/pytorch_lightning/metrics/__init__.py:43: LightningDeprecationWarning: `pytorch_lightning.metrics.*` module has been renamed to `torchmetrics.*` and split off to its own package (https://github.com/PyTorchLightning/metrics) since v1.3 and will be removed in v1.5
  rank_zero_deprecation(
/Users/eubin/Desktop/Projects/Big/wisdomify/wisdomifyenv/lib/python3.8/site-packages/pytorch_lightning/metrics/__init__.py:43: LightningDeprecationWarning: `pytorch_lightning.metrics.*` module has been renamed to `torchmetrics.*` and split off to its own package (https://github.com/PyTorchLightning/metrics) since v1.3 and will be removed in v1.5
  rank_zero_deprecation(
/Users/eubin/Desktop/Projects/Big/wisdomify/wisdomifyenv/lib/python3.8/site-packages/pytorch_lightning/metrics/__init__.py:43: LightningDeprecationWarning: `pytorch_lightning.metrics.*` module has been renamed to `torchmetrics.*` and split off to its own package (https://github.com/PyTorchLightning/metrics) since v1.3 and will be removed in v1.5
  rank_zero_deprecation(
batch:0 (0.0, 0.0, 1.0, 1.0, 1.0)
batch:1 (0.0, 0.0, 1.0, 1.0, 1.0)
### final ###
data: wisdom2def
median: 0.0
var: 0.0
top1: 1.0
top10: 1.0
top100: 1.0

```


내 생각에 문제는 여기와 관련이 있을 것으로 보임. "fork" 하는 부분이 여기인 것 같다.
https://github.com/eubinecto/wisdomify/blob/d457e61cc47633f11add94c24ef5c2b69dd2dad8/wisdomify/main/eval.py#L55

**Comment by ArtemisDicoTiar** [comment]
메인 실행에 아래 코드를 넣어서 환경변수를 설정하면 없어지긴한다.
~~~python
import os
os.environ["TOKENIZERS_PARALLELISM"] = "false"
~~~
문제는 이러면 parallelism이 비활성화된다.

tokenizer를 포크 전에 안쓰면 될거 같은데 어떻게 수정하면 되려나

**Comment by ArtemisDicoTiar** [comment]
오 그냥 저 설정을 true로 놔도 사라진다. 아마도 true로 둔 경우 parallelism으로 실행 못하는 경우 알아서 병렬처리를 없애는 거 같다.

**Comment by ArtemisDicoTiar** [comment]
main 내에 있는 모든 실행에 대해서 저 워닝 이슈 발생을 막기 위해
~~~
wisdomify/main/__init__.py
~~~
에 설정을 추가함.

**Comment by ArtemisDicoTiar** [comment]
#23 PR에서 얘기한 것처럼 토크나이저를 데이터셋에서 따로 빼봤으나 해결되지 않음. (클래스간 상호작용을 내가 정확히 몰라서 그럴수도 있음.)

좀더 이리저리 코드를 조작해보고 HF/tokenizers 리포를 찾아본 결과

해당 워닝 발생은 multiprocessing이 설정되어 있어서였고. (num_workers >0)
[HF/tokenizers](https://github.com/huggingface/tokenizers/commit/ae743f5dc15d0440ce43acff1b7f9a41c1c4579c)
리포를 찾아보니 
해당 워닝 발생이유는 다음과 같다.

~~~rust
// For users using multiprocessing in python, it is quite easy to fork the process running
// tokenizers, ending up with a deadlock because we internaly make use of multithreading. So
// we register a callback to be called in the event of a fork so that we can warn the user.
~~~

해결 방안은 다음과 같다.
~~~python
dataloader = DataLoader(dataset, batch_size, shuffle, num_workers=0)
~~~

멀티프로세싱을 파이썬(토치)에서 사용하지 않게 막는 거다.
실제로 실행해보니 약 4~5초에서 시간이 덜 걸렸다. (즉, 멀티프로세싱을 걸었을 때 포크 땜에 생기는 데드락을 방지하는 코드를 돌리는 데에 시간이 소요되는 거 같다.)

데이터 로더를 병렬처리 할 수 있게 하는 방법을 좀더 찾아보면 괜찮은 해결책을 찾을 수 있을 거 같다.
